### PR TITLE
[IOS] Add smartInsertDelete prop to TextInput component

### PR DESCRIPTION
**Pull request by fabioh8010**
# Upstream PR Link

https://github.com/facebook/react-native/pull/36111

## Summary

This PR add support to configure the smartInsertDeleteType property in iOS inputs [as described here](https://developer.apple.com/documentation/uikit/uitextinputtraits/2865828-smartinsertdeletetype), making possible toggle this autoformatting behavior.

This will allow fixing this issue: https://github.com/Expensify/App/issues/14049

## Changelog

[IOS] [ADDED] - Add smartInsertDelete prop to TextInput component

## Test Plan

* Added an example in the RNTester app to test this behavior in iOS.
* If the `smartInsertDelete` prop is `true` or `undefined`, the system will use the default autoformatting behavior.
* If the `smartInsertDelete` prop is `false`, the system will disable the autoformatting behavior.

https://user-images.githubusercontent.com/20051562/217862828-70c20344-d687-4824-8f5d-d591eff856ef.mp4

### CODE REVIEW

**Comment by fabioh8010** [comment]
@sketchydroide @mananjadhav @lschurr

**Review by mountiny — APPROVED** [review]
code is looking good to me although I dont have much experience with this code.

File: React/Fabric/Mounting/ComponentViews/TextInput/RCTTextInputComponentView.mm [hunk 1]
```diff
@@ -185,6 +185,13 @@ - (void)updateProps:(Props::Shared const &)props oldProps:(Props::Shared const &
     }
   }
 
+  if (newTextInputProps.traits.smartInsertDelete != oldTextInputProps.traits.smartInsertDelete) {
```
Comment (sketchydroide) [inline React/Fabric/Mounting/ComponentViews/TextInput/RCTTextInputComponentView.mm:196]: Can you add a comment on what this is doing specifically

File: React/Fabric/Mounting/ComponentViews/TextInput/RCTTextInputUtils.h [hunk 2]
```diff
@@ -41,4 +41,7 @@ UITextContentType RCTUITextContentTypeFromString(std::string const &contentType)
 API_AVAILABLE(ios(12.0))
 UITextInputPasswordRules *RCTUITextInputPasswordRulesFromString(std::string const &passwordRules);
 
+API_AVAILABLE(ios(11.0))
```
Comment (sketchydroide) [inline React/Fabric/Mounting/ComponentViews/TextInput/RCTTextInputUtils.h:44]: I guess this is for the best but I don't think we even build for this iOS version :)

**Review by sketchydroide — APPROVED** [review]
LGTM
Nothing iOS wise catches my eye as wrong, let's see if this tests good

**Comment by sketchydroide** [comment]
@mananjadhav can you see this?

File: React/Fabric/Mounting/ComponentViews/TextInput/RCTTextInputComponentView.mm (hunk 1 above)
Comment (fabioh8010) [reply React/Fabric/Mounting/ComponentViews/TextInput/RCTTextInputComponentView.mm:196]: This method is called when updating component's props, so for each prop we compare if the new value is different from the old one. If it is different, we update the native view with the new value.

`RCTUITextSmartInsertDeleteTypeFromOptionalBool` is used to convert the boolean value coming from JS side to the appropriate `UITextSmartInsertDeleteType` value, that is required for this `smartInsertDeleteType` attribute.

File: React/Fabric/Mounting/ComponentViews/TextInput/RCTTextInputUtils.h (hunk 2 above)
Comment (fabioh8010) [reply React/Fabric/Mounting/ComponentViews/TextInput/RCTTextInputUtils.h:44]: Yeah, it's just to follow the pattern :)

File: React/Fabric/Mounting/ComponentViews/TextInput/RCTTextInputComponentView.mm (hunk 1 above)
Comment (sketchydroide) [reply React/Fabric/Mounting/ComponentViews/TextInput/RCTTextInputComponentView.mm:196]: I meant in the code :D

Comment (fabioh8010) [reply React/Fabric/Mounting/ComponentViews/TextInput/RCTTextInputComponentView.mm:196]: Oh ok, sure 😂

Comment (fabioh8010) [reply React/Fabric/Mounting/ComponentViews/TextInput/RCTTextInputComponentView.mm:196]: Done.

**Review by roryabraham — CHANGES_REQUESTED** [review]
Sorry for the inconvenience, but this PR needs to be updated to target the `Expensify-0.71.2-alpha.0` branch

**Comment by fabioh8010** [comment]
Hi @roryabraham, could you please review the changes?

**Review by mananjadhav — APPROVED** [review]

**Comment by mananjadhav** [comment]
@sketchydroide I just reviewed this today.

**Review by roryabraham — APPROVED** [review]
Seems fine to me

**Pull request merged**